api/README.md

```diff
@@ -0,0 +1,69 @@
+# API Development
+
+This documentation is for developers to extend the current APIs. For the
+administrative documentation on the APIs, see [SETUP/API.md](SETUP/API.md).
+
+## Design
+
+The DP API uses a RESTful design, where URLs uniquely access resources.
+
+### Routing work to functions
+
+The framework is designed for a single function to handle all actions for
+a specific portion of the URI. If there are additional pieces to the URI,
+the `api_router()` is called to handle the remainder. This is achieved
+by using well-defined function names that reflect the URI being handled.
+
+Take the following URI:
+```
+/api/v1/projects/<projectid>/wordlists/<good>
+```
+
+After handling authentication and rate limiting, the router dispatches
+the initial work to `api_v1()`. That function does not consume part of the
+URI or query parameters so it calls back to the router. The router then
+dispatches the remaining work to `api_v1_projects()`.
+
+`api_v1_projects()` checks the URI to see if there are additional parts on
+the URI. If there are, it loads the `projectid` off the URI, validates it
+and passes the Project object back to the router for the next handler.
+
+The router then calls `api_v1_projects_PROJECTID_wordlists()` which takes the
+project, determines which words list to use (`good`) and based on the method
+type either returns the word list (GET) or updates the word list (PUT). In
+both cases it returns a data structure that will be output to the user in
+JSON format.
```

Comment:
A thought for future development here: We might want to be able to associate API keys with ACLs which only allow particular ACLs, and perhaps allow a user to generate multiple API keys each with different ACL sets.
Specifically, it would be good to allow users to read but not write the wordlist.

Reply:
Conceptually, I totally agree with you. In practice, the object authorization model in the code is completely decentralized right now, which makes this almost impossible. Examples:
* The project page limits the user's ability to see the clearance line based on who they are, but the Project object doesn't know anything about such limitations.
* The user's ability to edit the project word lists is embedded in the word list editing page not in the underlying code that saves the word lists.
Both of the above determine this by the currently-logged-in user (`$pguser`).

The word list example is a good one -- I mis-coded the enforcement in the API as it currently doesn't match the rest of the site as all users can see word lists.

I think having a 1:1 mapping for users is the least-bad option right now to get a working API. We have a fair amount of work to do before we can break that into 1:many with different authorization levels.

Reply:
Anterior Cruciate Ligament?

Reply:
In this case, [Access Control Lists](https://en.wikipedia.org/wiki/Access-control_list).